src/machine/parsed_results.rs

```diff
@@ -1,144 +1,485 @@
+            }
+
+            fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                Ok(Value::Integer(v.into()))
+            }
+
+            fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                Ok(Value::Integer(v.into()))
+            }
+
+            fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                Ok(Value::String(v))
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                self.visit_string::<E>(v.into())
+            }
+
+            fn visit_f64<E>(self, v: f64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                Ok(Value::Float(OrderedFloat(v)))
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::SeqAccess<'de>,
+            {
+                let mut buf = if let Some(cap) = seq.size_hint() {
+                    Vec::with_capacity(cap)
+                } else {
+                    vec![]
+                };
+
+                while let Some(elem) = seq.next_element()? {
+                    buf.push(elem)
+                }
+
+                Ok(Value::List(buf))
+            }
+
+            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::MapAccess<'de>,
+            {
+                use serde::de::Error;
+
+                let first_key = map
+                    .next_key::<String>()?
+                    .ok_or_else(|| A::Error::invalid_length(0, &"some non-empty untagged value"))?;
+
+                match first_key.as_str() {
+                    "atom" => {
+                        let atom = map.next_value::<String>()?;
+                        Ok(Value::Atom(atom))
+                    }
+                    "variable" => {
+                        let var = map.next_value::<String>()?;
+                        Ok(Value::Var(var))
+                    }
+                    "integer" => {
+                        let int = map.next_value::<String>()?;
+                        let int = IBig::from_str(&int).map_err(|err| {
+                            A::Error::invalid_value(
+                                serde::de::Unexpected::Other(&err.to_string()),
+                                &"integer",
+                            )
+                        })?;
+                        Ok(Value::Integer(int))
+                    }
+                    "rational" => {
+                        let parts = map.next_value::<Vec<serde_json::Value>>()?;
```

Comment:
Can you think of any alternative? We need to store the numerator and denominator separately anyway, and being able to represent large integers is desirable.

Reply:
If we go with internally tagged we could just have a `numerator` and `divisor` field next to the `type` tag field i.e. `{ "type": "rational", numerator: ..., divisor: ... }` or with untagged the value could be an object/map rather than a list i.e. `{ "rational": { "numerator": ..., "divisor": ... }}`

Reply:
Oh, I like the internal tagging one.